content/blog/running-self-hosted-agents-in-azure-devops-using-an-azure-ad-access-token.md

```diff
@@ -0,0 +1,123 @@
+
+{{<img src="/img/blog/create-app-registration.png" class="img-fluid" title="Create app registration" >}}
+
+Next up, go to API Permissions and add the Azure Devops user_impersonation Delegated permission. This will allow the Service Principal to call the Azure Devops REST API, which is what the Agent software is doing when registering itself as a new agent inside a given agent pool.
+
+{{<img src="/img/blog/add-api-permissions-to-app-registration.png" class="img-fluid" title="Assign API Permissions" >}}
+
+Afterwards, we need to assign a role to this app registration so that it is bound to a specific subscription. Go to your given subscription and go to Identitiy and Access Management (IAM) in the blade on your left. We're going to add a new Role Assignment, so follow the given dialogue and make sure that you're setting our new SP to the "Reader" role.
+
+{{<img src="/img/blog/add-role-assignment.png" class="img-fluid" title="Add Role Assignment" >}}
+
+If you forgot to do this, during login you'll get the error "No subscriptions found for {ClientID}", hopefully pointing you towards fixing this by following the above.
+
+{{<img src="/img/blog/no-role-assignment-error.png" class="img-fluid" title="No Role Assignment" >}}
+
+Of course, just giving the Service Principal permission to call the ADO API is not enough to do what we want to do here, we also have to give the SP the right permissions inside ADO.
+
+First of all, navigate to the Users tab for your organization and add the service principal as a user with Basic permissions. We're going to make it a Project Contributor on the Azure Devops Project I've created earlier, conveniently named "SelfHostedAgent".
+
+{{<img src="/img/blog/add-sp-as-ado-user.png" class="img-fluid" title="Add Service Principal as ADO user" >}}
+
+Next up, we're going to navigate to the Agent Pool we've already created and make sure that the Service Principal (which is now treated the same as an ADO user) can actually Read & Manage the agent pool. For this, we need to add the User Permissions at the Organization Level and add our Service Principal as an Administrator.
+
+Why at the Organization level? If you set the Service Principal as an Administrator at the Project level, the agent will start up but it will give you an error message that somewhat resembles "{Service Principal} needs Manage permissions for pool {Pool Name} to perform the action" and then just error out. This probably has something to do with the fact that Agent Pools are actually an organization-level construct in ADO that are just shared between different projects based on permissions.
+
+{{<img src="/img/blog/set-agent-pool-user-permissions.png" class="img-fluid" title="Set Agent Pool user permissions" >}}
+
+After this you should be set! To be sure, let's just run a little script to make sure that we've got the basics down so far.
+
+```bash
+#/bin/bash
```

Comment:
I'm going to nitpick on your bash here 😛 

1. Avoid using `#!/bin/bash` if possible. Bash might not be located there (FreeBSD), or it may point to a painfully outdated version of Bash (MacOS). Instead, use `#!/usr/bin/env bash` which will use whatever is your preferred Bash (based on `$PATH`). I use Bash 5.2 via Homebrew, so my Bash is located in `/opt/homebrew/bin/bash` but my scripts don't need to know that.
2. For added safety, set the following options at the beginning of your script (by default!): `set -euo pipefail`. This is actually a combined setting for 3 options: `set -e` (exit if something errors), `set -u` (don't allow unset variables), and `set -o pipefail` (don't allow piped commands to fail either). Setting this makes your shell scripts _way_ more robust. 
3. make your constants actually constant. So instead of `FOO="bar"` use `readonly FOO="bar"`

Using these improvements, whenever the `az` command to set `TOKEN` fails, the script will just stop, as failures aren't allowed, and empty vars aren't allowed either. So even if `az` doesn't return a non-zero response, but also doesn't produce a token, your script will stop. It will also somewhat descriptively tell you why 😉 

- 

-

Reply:
Yes, that seems a lot better. Thanks!